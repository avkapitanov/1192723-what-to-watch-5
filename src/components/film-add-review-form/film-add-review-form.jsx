import React from "react";
import RatingStar from "../rating-star/rating-star";
import {REVIEW_RATING_MODIFIER, MAX_RATING_VALUE_FORM} from "../../const";
import PropTypes from "prop-types";
import withFilmAddReviewForm from "../../hocks/with-film-add-review-form/with-film-add-review-form";
import filmProp from "../film-page/film.prop";
import {fetchReview} from "../../store/api-actions";
import {connect} from "react-redux";
import FilmAddReviewFormErrorBlock from "../film-add-review-form-error-block/film-add-review-form-error-block";
import FilmAddReviewFormTextFieldNotification
  from "../film-add-review-form-text-field-notification/film-add-review-form-text-field-notification";

const FilmAddReviewForm = (props) => {
  const {onFormChange, onSubmitForm, film, rating, reviewText, hasTextError, setSubmitProcess, isDisabled, onError, hasError} = props;

  return (
    <div className="add-review">
      <form action="#" className="add-review__form" onChange={onFormChange} onSubmit={(evt) => {
        evt.preventDefault();
        setSubmitProcess();
        onSubmitForm(film.id, rating, reviewText, onError);
      }}>
        <RatingStar currentRating={rating} isDisabled={isDisabled} maxRating={MAX_RATING_VALUE_FORM}/>
        <div className="add-review__text">
          <textarea className="add-review__textarea" name="reviewText" id="review-text" placeholder="Review text" disabled={isDisabled}></textarea>
          <div className="add-review__submit">
            <FilmAddReviewFormTextFieldNotification textLength={reviewText.length} />
            <button className="add-review__btn" type="submit" disabled={hasTextError || isDisabled}>Post</button>
          </div>
        </div>
      </form>
      <FilmAddReviewFormErrorBlock hasError={hasError} />
    </div>
  );
};

FilmAddReviewForm.propTypes = {
  onFormChange: PropTypes.func.isRequired,
  onSubmitForm: PropTypes.func.isRequired,
  film: filmProp,
  rating: PropTypes.string.isRequired,
  reviewText: PropTypes.string.isRequired,
  hasTextError: PropTypes.bool.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  hasError: PropTypes.bool.isRequired,
  setSubmitProcess: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch) => ({
  onSubmitForm(id, rating, comment, callback) {
    dispatch(fetchReview(id, rating * REVIEW_RATING_MODIFIER, comment, callback));
  }
});

export default connect(null, mapDispatchToProps)(withFilmAddReviewForm(FilmAddReviewForm));
