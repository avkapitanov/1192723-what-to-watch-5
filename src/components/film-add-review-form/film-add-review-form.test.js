import React from "react";
import renderer from "react-test-renderer";
import {FilmAddReviewForm} from "./film-add-review-form";

const film = {
  id: 1,
  title: `We need to talk about Kevin`,
  genre: [`Drama`],
  year: 2011,
  poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/We_need_to_talk_about_Kevin.jpg`,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/We_need_to_talk_about_Kevin.jpg`,
  background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/We_need_to_talk_about_Kevin.jpg`,
  video: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  backgroundColor: `#E1DFDE`,
  rating: 3.8,
  scoresCount: 123240,
  description: `Kevin's mother struggles to love her strange child, despite the increasingly dangerous things he says and does as he grows up. But Kevin is just getting started, and his final act will be beyond anything anyone imagined.`,
  director: `Lynne Ramsay`,
  starring: [
    `Tilda Swinton`,
    `John C. Reilly`,
    `Ezra Miller`
  ],
  runTime: 112,
  isFavorite: false
};

const noop = () => {};

describe(`FilmAddReviewForm component render correctly`, () => {
  it(`Init state`, () => {
    const tree = renderer.create(
        <FilmAddReviewForm
          film={film}
          onFormChange={noop}
          onSubmitForm={noop}
          rating={`3`}
          reviewText={``}
          hasTextError={false}
          isDisabled={false}
          hasError={false}
          setSubmitProcess={noop}
          onError={noop}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`With error`, () => {
    const tree = renderer.create(
        <FilmAddReviewForm
          film={film}
          onFormChange={noop}
          onSubmitForm={noop}
          rating={`3`}
          reviewText={``}
          hasTextError={false}
          isDisabled={false}
          hasError={true}
          setSubmitProcess={noop}
          onError={noop}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`With text length error`, () => {
    const tree = renderer.create(
        <FilmAddReviewForm
          film={film}
          onFormChange={noop}
          onSubmitForm={noop}
          rating={`3`}
          reviewText={``}
          hasTextError={true}
          isDisabled={false}
          hasError={false}
          setSubmitProcess={noop}
          onError={noop}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});


