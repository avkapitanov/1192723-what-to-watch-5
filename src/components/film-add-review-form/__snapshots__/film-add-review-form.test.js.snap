// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmAddReviewForm component render correctly Init state 1`] = `
<div
  className="add-review"
>
  <form
    action="#"
    className="add-review__form"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="rating"
    >
      <div
        className="rating__stars"
      >
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-1"
          name="rating"
          type="radio"
          value={1}
        />
        <label
          className="rating__label"
          htmlFor="star-1"
        >
          Rating 
          1
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-2"
          name="rating"
          type="radio"
          value={2}
        />
        <label
          className="rating__label"
          htmlFor="star-2"
        >
          Rating 
          2
        </label>
        <input
          className="rating__input"
          defaultChecked={true}
          disabled={false}
          id="star-3"
          name="rating"
          type="radio"
          value={3}
        />
        <label
          className="rating__label"
          htmlFor="star-3"
        >
          Rating 
          3
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-4"
          name="rating"
          type="radio"
          value={4}
        />
        <label
          className="rating__label"
          htmlFor="star-4"
        >
          Rating 
          4
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-5"
          name="rating"
          type="radio"
          value={5}
        />
        <label
          className="rating__label"
          htmlFor="star-5"
        >
          Rating 
          5
        </label>
      </div>
    </div>
    <div
      className="add-review__text"
    >
      <textarea
        className="add-review__textarea"
        disabled={false}
        id="review-text"
        name="reviewText"
        placeholder="Review text"
      />
      <div
        className="add-review__submit"
      >
        <div
          style={
            Object {
              "color": "red",
              "marginRight": "auto",
            }
          }
        >
          Min symbols 50. Current: 0
        </div>
        <button
          className="add-review__btn"
          disabled={false}
          type="submit"
        >
          Post
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`FilmAddReviewForm component render correctly With error 1`] = `
<div
  className="add-review"
>
  <form
    action="#"
    className="add-review__form"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="rating"
    >
      <div
        className="rating__stars"
      >
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-1"
          name="rating"
          type="radio"
          value={1}
        />
        <label
          className="rating__label"
          htmlFor="star-1"
        >
          Rating 
          1
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-2"
          name="rating"
          type="radio"
          value={2}
        />
        <label
          className="rating__label"
          htmlFor="star-2"
        >
          Rating 
          2
        </label>
        <input
          className="rating__input"
          defaultChecked={true}
          disabled={false}
          id="star-3"
          name="rating"
          type="radio"
          value={3}
        />
        <label
          className="rating__label"
          htmlFor="star-3"
        >
          Rating 
          3
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-4"
          name="rating"
          type="radio"
          value={4}
        />
        <label
          className="rating__label"
          htmlFor="star-4"
        >
          Rating 
          4
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-5"
          name="rating"
          type="radio"
          value={5}
        />
        <label
          className="rating__label"
          htmlFor="star-5"
        >
          Rating 
          5
        </label>
      </div>
    </div>
    <div
      className="add-review__text"
    >
      <textarea
        className="add-review__textarea"
        disabled={false}
        id="review-text"
        name="reviewText"
        placeholder="Review text"
      />
      <div
        className="add-review__submit"
      >
        <div
          style={
            Object {
              "color": "red",
              "marginRight": "auto",
            }
          }
        >
          Min symbols 50. Current: 0
        </div>
        <button
          className="add-review__btn"
          disabled={false}
          type="submit"
        >
          Post
        </button>
      </div>
    </div>
  </form>
  <div
    style={
      Object {
        "color": "red",
        "paddingTop": "10px",
        "textAlign": "center",
      }
    }
  >
    Something wrong, please, try again later
  </div>
</div>
`;

exports[`FilmAddReviewForm component render correctly With text length error 1`] = `
<div
  className="add-review"
>
  <form
    action="#"
    className="add-review__form"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="rating"
    >
      <div
        className="rating__stars"
      >
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-1"
          name="rating"
          type="radio"
          value={1}
        />
        <label
          className="rating__label"
          htmlFor="star-1"
        >
          Rating 
          1
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-2"
          name="rating"
          type="radio"
          value={2}
        />
        <label
          className="rating__label"
          htmlFor="star-2"
        >
          Rating 
          2
        </label>
        <input
          className="rating__input"
          defaultChecked={true}
          disabled={false}
          id="star-3"
          name="rating"
          type="radio"
          value={3}
        />
        <label
          className="rating__label"
          htmlFor="star-3"
        >
          Rating 
          3
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-4"
          name="rating"
          type="radio"
          value={4}
        />
        <label
          className="rating__label"
          htmlFor="star-4"
        >
          Rating 
          4
        </label>
        <input
          className="rating__input"
          defaultChecked={false}
          disabled={false}
          id="star-5"
          name="rating"
          type="radio"
          value={5}
        />
        <label
          className="rating__label"
          htmlFor="star-5"
        >
          Rating 
          5
        </label>
      </div>
    </div>
    <div
      className="add-review__text"
    >
      <textarea
        className="add-review__textarea"
        disabled={false}
        id="review-text"
        name="reviewText"
        placeholder="Review text"
      />
      <div
        className="add-review__submit"
      >
        <div
          style={
            Object {
              "color": "red",
              "marginRight": "auto",
            }
          }
        >
          Min symbols 50. Current: 0
        </div>
        <button
          className="add-review__btn"
          disabled={true}
          type="submit"
        >
          Post
        </button>
      </div>
    </div>
  </form>
</div>
`;
